### DISCUSSION TITLE
Support for user "groups" or "organisations"

### DISCUSSION THREAD

**Discussion by pigfrown** [ory/kratos#2057] [Q&A] [answered]
Hello,

I am attempting to implement an Organisation/User hierarchy using Kratos for authentication and have a few questions.

I need to have concepts of "Organisations", "Org Admins", and "Org Users". An Organisation + Org Admin would be created by myself manually, and then an Org Admin would be able to use the Kratos API to create new users underneath their Organisation.

After reading through the docs/discussion I gather there is no way to support this out of the box with Kratos? Originally I planned on using traits to store a users Organisation, but after reading that traits can always be modified by the user I abandoned that approach.

My plan now is to a separate service that maintains a database that tracks all organisations and associates Kratos user ID's with orgs. I would also use this database to associate users with roles that would be used by OPA to authorize access to particular services/resources.

Does this approach make sense? Is there another way I could achieve this using just Kratos?

Thanks

**Comment by vinckr** [comment, marked as answer]
Hello pigfrown 👋 
The trait issue is actually something we want to change: https://github.com/ory/kratos/issues/117, there has also been some movement recently, see the linked https://github.com/ory/kratos/issues/1777
So technically it would be possible with traits then.
Generally we recommend to use [Ory Keto](https://www.ory.sh/keto/) to implement things like roles/organizations, but OPA works too (the previous versions of Ory Keto are also based on OPA).
See also this [video about Keto](https://www.youtube.com/watch?v=lGRMYkQrNb0) and [the example](https://www.ory.sh/keto/docs/examples/olymp-file-sharing).
Let me know if that answered your question.

**Reply by pigfrown** [reply]
Hi Vinckr

Keto looks interesting and could be a good fit for us. The main reason we were favouring OPA was the ability to run as a sidecar next to each of our services. In this way services would not be required to query an external server for authz (keto?), but instead ask the OPA instance running as a sidecar in the same container. Perhaps we are prematurely optimising here though!

Can a similar "sidecar" approach be taken with keto? I looked at Oathkeeper but it still seems to rely on external calls to keto to make policy decisions?

Thanks!

**Reply by vinckr** [reply]
I am not an expert in container orchestration, but I think it should be possible to run Ory Keto in the same container as a sidecar! 
I am not sure if this is the easiest approach, and Ory Keto is designed for high-performance (think Google-level traffic) so latency should not be a problem if it runs by itself.
Maybe the Ory Keto wizard 🧙 @zepatrik can confirm this?

**Reply by zepatrik** [reply]
The more common way of deploying Keto seems to be as a separate service because it scales independently, but I guess there is nothing preventing a sidecar deployment. Best would be to just try and compare. For a reduced latency we recommend the gRPC APIs over REST, but both work well.

**Reply by pigfrown** [reply]
Interesting!

I'll experiment with modelling our org/user hierarchy in keto. I think I might be worrying about performance a little too soon.

Thanks for the input

**Reply by vinckr** [reply]
Patrik also held a workshop about designing authorization with Ory Keto, check it out: 
https://www.youtube.com/watch?v=lGRMYkQrNb0
There is also a github repo with the examples linked.
